File: c/scrabble_score.c
#include "scrabble_score.h"

#include <ctype.h>

unsigned int score(const char *word) {
  int res = 0;

  while (*word != '\0') {
    int sc = 1;
    switch (tolower(*word)) {
    case 'd':
    case 'g':
      sc = 2;
      break;
    case 'b':
    case 'c':
    case 'm':
    case 'p':
      sc = 3;
      break;
    case 'f':
    case 'h':
    case 'v':
    case 'w':
    case 'y':
      sc = 4;
      break;
    case 'k':
      sc = 5;
      break;
    case 'j':
    case 'x':
      sc = 8;
      break;
    case 'q':
    case 'z':
      sc = 10;
      break;
    }
    res += sc;
    word++;
  }
  return res;
}

File: go/scrabble_score.go
package scrabble

import (
	"strings"
	"unicode"
)

var scoreTable = map[string]int{
	"aeioulnrst": 1,
	"dg":         2,
	"bcmp":       3,
	"fhvwy":      4,
	"k":          5,
	"jx":         8,
	"qz":         10,
}

func point(r rune) int {
	l := unicode.ToLower(r)

	for s, p := range scoreTable {
		if strings.ContainsRune(s, l) {
			return p
		}
	}

	return 0
}

// Score Given a word, compute the scrabble score for that word.
func Score(word string) (sum int) {
	for _, ch := range word {
		sum += point(ch)
	}
	return
}
